api/v1beta1/azuremanagedcontrolplane_webhook.go

```diff
@@ -657,10 +736,85 @@ func (m *AzureManagedControlPlane) validateAADProfileUpdateAndLocalAccounts(old
+			allErrs = append(allErrs, errAzureKeyVaultKms...)
+		}
+		if errWorkloadIdentity := m.validateWorkloadIdentityUpdate(old); errWorkloadIdentity != nil {
+			allErrs = append(allErrs, errWorkloadIdentity...)
+		}
+		if errWorkloadIdentity := m.validateImageCleanerUpdate(old); errWorkloadIdentity != nil {
+			allErrs = append(allErrs, errWorkloadIdentity...)
+		}
+		if errWorkloadIdentity := m.validateDefender(old); errWorkloadIdentity != nil {
+			allErrs = append(allErrs, errWorkloadIdentity...)
+		}
+	}
+	return allErrs
+}
+
+// validateAzureKeyVaultKmsUpdate validates AzureKeyVaultKmsUpdate profile.
+func (m *AzureManagedControlPlane) validateAzureKeyVaultKmsUpdate(old *AzureManagedControlPlane) field.ErrorList {
+	var allErrs field.ErrorList
+	if old.Spec.SecurityProfile.AzureKeyVaultKms != nil {
+		if m.Spec.SecurityProfile == nil || m.Spec.SecurityProfile.AzureKeyVaultKms == nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("Spec", "SecurityProfile", "AzureKeyVaultKms"),
+				nil, "cannot unset Spec.SecurityProfile.AzureKeyVaultKms profile"))
+			return allErrs
+		}
+		if err := webhookutils.ValidateImmutable(
+			field.NewPath("Spec", "SecurityProfile", "AzureKeyVaultKms"),
+			old.Spec.SecurityProfile.AzureKeyVaultKms,
+			m.Spec.SecurityProfile.AzureKeyVaultKms,
+		); err != nil {
+			allErrs = append(allErrs, err)
+		}
+	}
+	return allErrs
+}
+
+// validateWorkloadIdentityUpdate validates WorkloadIdentityUpdate profile.
+func (m *AzureManagedControlPlane) validateWorkloadIdentityUpdate(old *AzureManagedControlPlane) field.ErrorList {
+	var allErrs field.ErrorList
+	if old.Spec.SecurityProfile.WorkloadIdentity != nil {
+		if m.Spec.SecurityProfile == nil || m.Spec.SecurityProfile.WorkloadIdentity == nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("Spec", "SecurityProfile", "WorkloadIdentity"),
+				nil, "cannot unset Spec.SecurityProfile.WorkloadIdentity"))
+		}
+	}
+	return allErrs
+}
+
+// validateImageCleanerUpdate validates ImageCleanerUpdate profile.
+func (m *AzureManagedControlPlane) validateImageCleanerUpdate(old *AzureManagedControlPlane) field.ErrorList {
+	var allErrs field.ErrorList
+	if old.Spec.SecurityProfile.ImageCleaner != nil {
+		if m.Spec.SecurityProfile == nil || m.Spec.SecurityProfile.ImageCleaner == nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("Spec", "SecurityProfile", "WorkloadIdentity"),
+				nil, "cannot unset Spec.SecurityProfile.validateImageCleaner"))
+		}
+	}
+	return allErrs
+}
+
+// validateDefender validates defender profile.
+func (m *AzureManagedControlPlane) validateDefender(old *AzureManagedControlPlane) field.ErrorList {
+	var allErrs field.ErrorList
+	if old.Spec.SecurityProfile.Defender != nil {
+		if m.Spec.SecurityProfile == nil || m.Spec.SecurityProfile.Defender == nil {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("Spec", "SecurityProfile", "WorkloadIdentity"),
```

Comment:
same question about disabling defender https://learn.microsoft.com/en-us/azure/defender-for-cloud/defender-for-containers-enable?tabs=aks-deploy-portal%2Ck8s-deploy-asc%2Ck8s-verify-asc%2Ck8s-remove-arc%2Caks-removeprofile-api&pivots=defender-for-container-aks#remove-the-defender-agent

Reply:
You are right these fields can be disabled but cannot be unset. We need to set it as false to disable once enabled.